File: sample-app-scala/app/data/ServiceClient.scala
package data

import com.ybrikman.ping.scalaapi.dedupe.DedupingCache
import play.api.libs.ws.{WSResponse, WSClient}
import play.api.libs.concurrent.Execution.Implicits._
import play.api.mvc.RequestHeader

import scala.concurrent.Future

class ServiceClient(ws: WSClient, cache: DedupingCache[String, Future[WSResponse]]) {

  def remoteCall(url: String)(implicit rh: RequestHeader): Future[WSResponse] = {
    cache.get(url, ws.url(url).get())
  }
}

File: sample-app-java/app/data/ServiceClient.java
package data;

import com.ybrikman.ping.javaapi.dedupe.DedupingCache;
import play.libs.F;
import play.libs.ws.WSClient;
import play.libs.ws.WSResponse;

import javax.inject.Inject;

public class ServiceClient {
  private final WSClient ws;
  private final DedupingCache<String, F.Promise<WSResponse>> cache;

  @Inject
  public ServiceClient(WSClient ws, DedupingCache<String, F.Promise<WSResponse>> cache) {
    this.ws = ws;
    this.cache = cache;
  }

  public F.Promise<WSResponse> remoteCall(String url) {
    return cache.get(url, () -> ws.url(url).get());
  }
}
